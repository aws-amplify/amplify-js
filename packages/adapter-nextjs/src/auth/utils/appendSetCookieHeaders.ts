// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { CookieStorage } from 'aws-amplify/adapter-core';

import { serializeCookie } from '../../utils/cookie';

export const appendSetCookieHeaders = (
	headers: Headers,
	cookies: { name: string; value: string }[],
	setCookieOptions?: CookieStorage.SetCookieOptions,
): void => {
	for (const { name, value } of cookies) {
		headers.append(
			'Set-Cookie',
			serializeCookie(name, value, setCookieOptions),
		);
	}
};
