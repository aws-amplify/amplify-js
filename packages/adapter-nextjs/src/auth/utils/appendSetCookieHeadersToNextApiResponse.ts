// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { NextApiResponse } from 'next';
import { CookieStorage } from 'aws-amplify/adapter-core';

import { serializeCookie } from '../../utils/cookie';

export const appendSetCookieHeadersToNextApiResponse = (
	response: NextApiResponse,
	cookies: { name: string; value: string }[],
	setCookieOptions?: CookieStorage.SetCookieOptions,
): void => {
	for (const { name, value } of cookies) {
		response.appendHeader(
			'Set-Cookie',
			serializeCookie(name, value, setCookieOptions),
		);
	}
};
