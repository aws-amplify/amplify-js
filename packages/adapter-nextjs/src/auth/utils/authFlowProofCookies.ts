// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CookieStorage } from 'aws-amplify/adapter-core';

import {
	AUTH_FLOW_PROOF_MAX_AGE,
	IS_SIGNING_OUT_COOKIE_NAME,
	IS_SIGNING_OUT_REDIRECTING_COOKIE_NAME,
	PKCE_COOKIE_NAME,
	REMOVE_COOKIE_MAX_AGE,
	STATE_COOKIE_NAME,
} from '../constant';

import { isSSLOrigin } from './origin';

export const createSignInFlowProofCookies = ({
	state,
	pkce,
}: {
	state: string;
	pkce: string;
}) => [
	{
		name: PKCE_COOKIE_NAME,
		value: pkce,
	},
	{
		name: STATE_COOKIE_NAME,
		value: state,
	},
];

export const createSignOutFlowProofCookies = () => [
	{
		name: IS_SIGNING_OUT_COOKIE_NAME,
		value: 'true',
	},
	{
		name: IS_SIGNING_OUT_REDIRECTING_COOKIE_NAME,
		value: 'true',
	},
];

export const createAuthFlowProofCookiesSetOptions = (
	setCookieOptions: CookieStorage.SetCookieOptions,
	origin: string,
) => ({
	domain: setCookieOptions?.domain,
	path: '/',
	httpOnly: true,
	secure: isSSLOrigin(origin),
	sameSite: 'lax' as const,
	maxAge: AUTH_FLOW_PROOF_MAX_AGE,
});

export const createAuthFlowProofCookiesRemoveOptions = (
	setCookieOptions: CookieStorage.SetCookieOptions,
) => ({
	domain: setCookieOptions?.domain,
	path: '/',
	maxAge: REMOVE_COOKIE_MAX_AGE,
});
