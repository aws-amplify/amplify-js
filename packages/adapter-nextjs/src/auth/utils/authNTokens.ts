// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { OAuthConfig } from '@aws-amplify/core';

import { OAUTH_GRANT_TYPE } from '../constant';
import { OAuthTokenExchangeResult, OAuthTokenRevocationResult } from '../types';

import {
	createUrlSearchParamsForTokenExchange,
	createUrlSearchParamsForTokenRevocation,
} from './createUrlSearchParams';
import {
	createRevokeEndpoint,
	createTokenEndpoint,
} from './cognitoHostedUIEndpoints';

export const exchangeAuthNTokens = async ({
	redirectUri,
	userPoolClientId,
	oAuthConfig,
	code,
	codeVerifier,
}: {
	redirectUri: string;
	userPoolClientId: string;
	oAuthConfig: OAuthConfig;
	code: string;
	codeVerifier: string;
}): Promise<OAuthTokenExchangeResult> => {
	const searchParams = createUrlSearchParamsForTokenExchange({
		client_id: userPoolClientId,
		code,
		redirect_uri: redirectUri,
		code_verifier: codeVerifier,
		grant_type: OAUTH_GRANT_TYPE,
	});

	const oAuthTokenEndpoint = createTokenEndpoint(oAuthConfig.domain);
	const tokenExchangeResponse = await fetch(oAuthTokenEndpoint, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/x-www-form-urlencoded',
			'Cache-Control': 'no-cache',
		},
		body: searchParams.toString(),
	});

	// Exchanging an authorization code grant with PKCE for tokens with
	// `grant_type=authorization_code` produces a stable shape of payload.
	// Details see https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html
	// Possible errors: invalid_request|invalid_client|invalid_grant|unauthorized_client|unsupported_grant_type
	// Should not happen unless configuration is wrong;
	return (await tokenExchangeResponse.json()) as OAuthTokenExchangeResult;
};

export const revokeAuthNTokens = async ({
	userPoolClientId,
	refreshToken,
	endpointDomain,
}: {
	userPoolClientId: string;
	refreshToken: string;
	endpointDomain: string;
}): Promise<OAuthTokenRevocationResult> => {
	const searchParams = createUrlSearchParamsForTokenRevocation({
		client_id: userPoolClientId,
		token: refreshToken,
	});
	const oAuthTokenRevocationEndpoint = createRevokeEndpoint(endpointDomain);
	const response = await fetch(oAuthTokenRevocationEndpoint, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/x-www-form-urlencoded',
			'Cache-Control': 'no-cache',
		},
		body: searchParams.toString(),
	});
	const contentLength = parseInt(
		response.headers.get('Content-Length') ?? '0',
		10,
	);

	return contentLength === 0
		? {}
		: ((await response.json()) as OAuthTokenRevocationResult);
};
