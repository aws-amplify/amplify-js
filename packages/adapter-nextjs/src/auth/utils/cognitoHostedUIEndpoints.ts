// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export const createAuthorizeEndpoint = (
	domain: string,
	urlSearchParams: URLSearchParams,
): string =>
	new URL(
		`https://${domain}/oauth2/authorize?${urlSearchParams.toString()}`,
	).toString();

export const createTokenEndpoint = (domain: string): string =>
	new URL(`https://${domain}/oauth2/token`).toString();

export const createRevokeEndpoint = (domain: string) =>
	new URL(`https://${domain}/oauth2/revoke`).toString();

export const createSignUpEndpoint = (
	domain: string,
	urlSearchParams: URLSearchParams,
): string =>
	new URL(`https://${domain}/signup?${urlSearchParams.toString()}`).toString();

export const createLogoutEndpoint = (
	domain: string,
	urlSearchParams: URLSearchParams,
): string =>
	new URL(`https://${domain}/logout?${urlSearchParams.toString()}`).toString();
