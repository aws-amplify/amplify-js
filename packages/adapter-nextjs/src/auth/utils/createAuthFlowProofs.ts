// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { urlSafeEncode } from '@aws-amplify/core/internals/utils';
import { generateCodeVerifier, generateState } from 'aws-amplify/adapter-core';

export const createAuthFlowProofs = ({
	customState,
}: {
	customState?: string;
}): {
	codeVerifier: ReturnType<typeof generateCodeVerifier>;
	state: string;
} => {
	const codeVerifier = generateCodeVerifier(128);
	const randomState = generateState();
	const state = customState
		? `${randomState}-${urlSafeEncode(customState)}`
		: randomState;

	return { codeVerifier, state };
};
