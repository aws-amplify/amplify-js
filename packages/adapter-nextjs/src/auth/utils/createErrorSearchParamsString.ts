// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export const createErrorSearchParamsString = ({
	error,
	errorDescription,
}: {
	error: string | null;
	errorDescription: string | null;
}): string => {
	const errorParams = new URLSearchParams();

	if (error) {
		errorParams.set('error', error);
	}

	if (errorDescription) {
		errorParams.set('error_description', errorDescription);
	}

	return errorParams.toString();
};
