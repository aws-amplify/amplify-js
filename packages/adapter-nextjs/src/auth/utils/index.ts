// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export {
	isAuthRoutesHandlersContext,
	isNextApiRequest,
	isNextApiResponse,
	isNextRequest,
} from './handlerParametersTypeAssertions';
export { isSupportedAuthApiRoutePath } from './isSupportedAuthApiRoutePath';
