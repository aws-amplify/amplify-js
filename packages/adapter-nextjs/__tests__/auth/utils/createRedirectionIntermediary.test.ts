import { createRedirectionIntermediary } from '../../../src/auth/utils/createRedirectionIntermediary';

describe('createOnSignInCompletedRedirectIntermediate', () => {
	it('returns html with script that redirects to the redirectUrl', () => {
		const redirectUrl = 'https://example.com';
		const result = createRedirectionIntermediary({
			redirectTo: redirectUrl,
		});

		expect(result).toMatchSnapshot();
	});
});
