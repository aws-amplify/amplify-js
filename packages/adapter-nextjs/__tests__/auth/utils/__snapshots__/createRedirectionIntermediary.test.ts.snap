// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOnSignInCompletedRedirectIntermediate returns html with script that redirects to the redirectUrl 1`] = `
"
<!DOCTYPE html>
	<html>
	<head>
			<title>Redirecting...</title>
			<meta http-equiv="refresh" content="0; URL='https://example.com'" />
			<script>window.location.replace("https://example.com")</script>
	</head>
	<body>
			<p>If you are not redirected automatically, follow this <a href="https://example.com">link to the new page</a>.</p>
	</body>
</html>"
`;
